pipeline/extract_demand.py

```diff
@@ -0,0 +1,245 @@
+import logging
+import os
+from datetime import datetime, timedelta, timezone
+from typing import Any, Dict, Optional, Tuple
+
+import boto3
+import pandas as pd
+import requests
+from dotenv import load_dotenv
+
+import config as cg
+from constants import Constants as ct
+
+# Global constants
+ENDPOINT = ct.DEMAND_ENDPOINT
+SAVE_NAME = ct.RAW_DEMAND_DATA_NAME
+SAVE_LOCATION = ct.RAW_DEMAND_DATA_PATH
+S3_BUCKET = ct.S3_BUCKET
+
+load_dotenv('.env')
+AWS_ACCESS_KEY = os.getenv('AWS_ACCESS_KEY')
+AWS_SECRET_KEY = os.getenv('AWS_SECRET_KEY')
+AWS_REGION = os.getenv('AWS_REGION')
+
+# Logging
+SCRIPT_NAME = os.path.basename(__file__).split(".")[0]
+LOGGING_LEVEL = logging.DEBUG
+logger = cg.setup_logging(SCRIPT_NAME, LOGGING_LEVEL)
+
+class APIClient:
+    """
+    Constructs the default parameters for a request, and makes the request
+    returning data over a specific range.
+    """
+
+    def __init__(self,
+                 base_url: str = ENDPOINT,
+                 logger: logging.Logger = logger) -> None:
+        """
+        Initialize class variables.
+        """
+        self.base_url = base_url
+        self.logger = logger
+
+    def construct_default_params(self) -> Dict[str, str]:
+        """
+        Get a time range. Currently fetches data from 12 hours ago to present.
+        """
+        twelve_hours_ago = datetime.now(timezone.utc) - timedelta(hours=12)
+        now = datetime.now(timezone.utc)
+
+        return {
+            'From': twelve_hours_ago.isoformat(),  # ISO 8601 format, already UTC-aware
+            'To': now.isoformat(),
+            'format': 'json'
+        }
+
+    def fetch_data(self) -> Optional[Dict[str, Any]]:
+        """
+        Uses the above-created time range to make an API request, returning data.
+        """
+        try:
+            response = requests.get(self.base_url, params=self.construct_default_params())
+            response.raise_for_status()
+            return response.json()
+        except requests.exceptions.RequestException as e:
+            self.logger.error(f"An error occurred: {e}")
+            return None
+
+
+class DataProcessor:
```

Comment:
Each extract script has a Dataprocessor class and APIclient class, we can remove the classes to keep the code DRY compliant